rfcs/015 - Reworked Compiler Front End.md

```diff
@@ -0,0 +1,179 @@
+There's a lot more complexity here!
+The current front-end would have to backtrack across all declarations, assignments, and Microstatement "rerefs" in order to ensure that the type assigned to `myMaybe` is `Maybe<int64>` once the `some(5)` assignment is reached.
+However, with the new front-end, the compiler would initially assign `myMaybe` to type `Maybe<T>`, then add a constraint that `T` can be assigned by the integer literal `5`.
+Then, once the `toInt8` call is reached, the compiler would add another constraint that for whatever type `T` is, there must be a function called `toInt8` that matches the signature `(T) -> int8`.
+If there are no other constraints, the compiler is able to resolve `T` as an `int64` before outputting the corresponding `amm`.
+
+The new compiler front-end will also be able to determine function signatures without requiring type annotations on parameters or the return type.
+A simple example might look like:
+```ln
+fn add4(num) {
+  return add(num, 4);
+}
+```
+
+At the time of writing this RFC, this function fails to parse due to a missing type annotation on `num`, but it's intended to support gradual typing even for function parameters in Alan.
+The current front-end will require substantial work to support this, requiring one of:
+- An interface (internal or in `root.ln`) that ensures there's an `add` operation for the type of `num` (for example, `Addable`).
+- Type-checking can only happen once the function is inlined.
+
+However, the new type resolution system would simply add a constraint that there be a function named `add` that satisfies the signature `(T, 4 as T) -> T` (ignoring the fact that arithmetic operations in Alan actually use `Result` values).
+This wouldn't require the `Addable` interface to be defined at all, and instead would allow for an anonymous interface that's only visible in the compiler to be created.
+The constraints would then not need to be computed for each time the function is called - instead, we must only ensure that the value passed in to `add4` satisfies the so-called "Addable" interface.
+
+A more complex example would be the `toString` implementation for `Maybe`, which could be rewritten as such under the new front-end:
+```ln
+fn toString(val) {
+  if val.isSome() {
+    return val.getMaybe().toString();
+  } else {
+    return 'none';
+  }
+}
+```
+
+In order to share more nuanced details about the new front-end, here are the steps the compiler might take upon having to use this function:
+1. Assign the variable `val` to type `T`. There are no constraints on the type `T`.
+2. Desugar the if/else. How this is done is unimportant, but the body of the `toString` function will effectively be desugared into (ignore the invalid `conditionTailFn` function body - it's required for `evalcond` to work but its contents won't be checked by the type checker, since this desugaring will actually happen just before generating the Microstatements):
+```ln
+const conditionalTable = newarr(2);
```

Comment:
Nit: while `const` does work alright now because of other bugs in the existing compiler, this should probably be a `let` here. :)